src/heuristic/HeuristicMapper.cpp

```diff
@@ -597,3 +615,162 @@ void HeuristicMapper::lookahead(const std::size_t      layer,
                                          // for better fidelity lookahead
   }
 }
+
+void HeuristicMapper::Node::applySWAP(const Edge& swap, Architecture& arch) {
+  nswaps++;
+  swaps.emplace_back();
+  const auto q1 = qubits.at(swap.first);
+  const auto q2 = qubits.at(swap.second);
+
+  qubits.at(swap.first)  = q2;
+  qubits.at(swap.second) = q1;
+
+  if (q1 != -1) {
+    locations.at(static_cast<std::size_t>(q1)) =
+        static_cast<std::int16_t>(swap.second);
+  }
+  if (q2 != -1) {
+    locations.at(static_cast<std::size_t>(q2)) =
+        static_cast<std::int16_t>(swap.first);
+  }
+
+  if (arch.getCouplingMap().find(swap) != arch.getCouplingMap().end() ||
+      arch.getCouplingMap().find(Edge{swap.second, swap.first}) !=
+          arch.getCouplingMap().end()) {
+    swaps.back().emplace_back(swap.first, swap.second, qc::SWAP);
+  } else {
+    throw QMAPException("Something wrong in applySWAP.");
+  }
+
+  if (arch.bidirectional()) {
+    costFixed += COST_BIDIRECTIONAL_SWAP;
```

Comment:
Just a side comment, because I thought about this now. This cost of a SWAP should probably be something that comes from the Architecture class. That class should be able to tell how a SWAP gate is realized on itself.
The Architecture class should probably also be able to answer the question "Can I apply this gate to these qubits?" and the follow-up question "How much does this cost me?" (Thinking of direction reversal costs).
Probably nothing for this PR, but good to keep in mind.

Reply:
I agree. Would be much cleaner to have some sort of getter function, with probably only negligible performance overhead.